plutus-core/plutus-core/src/PlutusCore/Value.hs

```diff
@@ -91,8 +104,48 @@ fromList =
 
 normalize :: NestedMap -> NestedMap
 normalize = Map.filter (not . Map.null) . Map.map (Map.filter (/= 0))
+{-# INLINEABLE normalize #-}
 
 instance Pretty Value where
   pretty = pretty . fmap (bimap toText (fmap (first toText))) . toList
    where
     toText = Text.decodeLatin1 . Base64.encode
+
+{-| \(O(\log \max(m, k))\), where \(m\) is the size of the outer map, and \(k\) is
+the size of the largest inner map.
+-}
+insertCoin :: ByteString -> ByteString -> Integer -> Value -> Value
+insertCoin currency token amt (Value outer sizes size) = case Map.lookup currency outer of
+  Nothing ->
+    let inner = Map.singleton token amt
+        outer' = Map.insert currency inner outer
+        sizes' = updateSizes 0 1 sizes
+        size' = size + 1
+     in Value outer' sizes' size'
+  Just inner ->
+    let exists = Map.member token inner
+        inner' = Map.insert token amt inner
+        outer' = Map.insert currency inner' outer
+        sizes' = if exists then sizes else updateSizes (Map.size inner) (Map.size inner') sizes
+        size' = if exists then size else size + 1
+     in Value outer' sizes' size'
+{-# INLINEABLE insertCoin #-}
+
+-- | \(O(n)\), where \(n\) is the total size of the maps.
+unionValue :: Value -> Value -> Value
+unionValue v v' = pack (Map.unionWith (Map.unionWith (+)) (unpack v) (unpack v'))
```

Comment:
See https://github.com/IntersectMBO/plutus/pull/7255/files#diff-0718f41c4fd914c8274438e1d1af10d48fe0e6fd738399c03a886f4ae85d474e. This was the best implementation of `union` I could get, maybe it can be modified to also take care of the sizes.

Reply:
In case the link doesn't work, the idea is to use `Map.merge`

Reply:
Updated using your implementation. 

> maybe it can be modified to also take care of the sizes. 

This seems prohibitively complex and error prone. I don't think it's worth doing so. Re-calculating the size only takes time proportional to the size of the outermaps, which is lower than the cost of union itself.